db/infinispan/src/main/java/org/commonjava/indy/infinispan/data/InfinispanStoreDataManager.java

```diff
@@ -0,0 +1,189 @@
+/**
+ * Copyright (C) 2011-2018 Red Hat, Inc. (https://github.com/Commonjava/indy)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.commonjava.indy.infinispan.data;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import org.commonjava.indy.audit.ChangeSummary;
+import org.commonjava.indy.data.IndyDataException;
+import org.commonjava.indy.data.NoOpStoreEventDispatcher;
+import org.commonjava.indy.data.StoreEventDispatcher;
+import org.commonjava.indy.db.common.AbstractStoreDataManager;
+import org.commonjava.indy.model.core.ArtifactStore;
+import org.commonjava.indy.model.core.StoreKey;
+import org.commonjava.indy.model.core.io.IndyObjectMapper;
+import org.commonjava.indy.subsys.infinispan.CacheHandle;
+import org.commonjava.indy.subsys.infinispan.CacheProducer;
+import org.infinispan.Cache;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.PostConstruct;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Alternative;
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static org.commonjava.indy.infinispan.data.StoreDataCacheProducer.STORE_DATA_CACHE;
+
+@ApplicationScoped
+@Alternative
+public class InfinispanStoreDataManager
+                extends AbstractStoreDataManager
+{
+    private final Logger logger = LoggerFactory.getLogger( getClass() );
+
+    @Inject
+    @StoreDataCache
+    private CacheHandle<String, String> stores;
```

Comment:
I'm wondering if it would be better to make ArtifactStore and descendants Externalizable. We might be able to annotate them for search indexing then, and make the ArtifactStoreQuery interface more efficient using the ISPN query stuff.

I know you mentioned something earlier about being able to go look at these definitions and debug based on directly viewing them. I think using a separate client that joins the cache cluster (or reads directly from the DB backing the cache, depending on whether it's a read-only client) might work just as well, and would allow us to avoid serializing / deserializing to JSON all the time.

Reply:
Ok. Will change to serializable Obj for values.

Reply:
@jdcasey After looking at the code, I found it is not easy to make a simple and nice externalization for ArtifactStore. It has three concrete subclasses, hosted, remote, group. And in between there is AbstractArtifactStore which is used by hosted and remote. We have to add externalization code in several places. The VERSION-ing is hard too. We have to add a version for each subclass. I am afraid this would introduce some complicated code - controlling the java serialized object versioning is not a simple task. On the other hand, using JSON text seems to eliminate all these troubles. How do you think?